cmd/whitesourceExecuteScan.go

```diff
@@ -0,0 +1,370 @@
+package cmd
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/SAP/jenkins-library/pkg/command"
+	"github.com/SAP/jenkins-library/pkg/log"
+	"github.com/SAP/jenkins-library/pkg/telemetry"
+	"github.com/SAP/jenkins-library/pkg/whitesource"
+	"golang.org/x/mod/modfile"
+)
+
+func whitesourceExecuteScan(config whitesourceExecuteScanOptions, telemetryData *telemetry.CustomData) {
+	// for command execution use Command
+	c := command.Command{}
+	sys := whitesource.NewSystem(config.ServiceURL, config.OrgToken, config.UserToken)
+	// reroute command output to logging framework
+	c.Stdout(log.Writer())
+	c.Stderr(log.Writer())
+
+	// error situations should stop execution through log.Entry().Fatal() call which leads to an os.Exit(1) in the end
+	err := runWhitesourceExecuteScan(config, sys, telemetryData, &c)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("step execution failed")
+	}
+}
+
+func runWhitesourceExecuteScan(config whitesourceExecuteScanOptions, sys whitesource.System, telemetryData *telemetry.CustomData, command *command.Command) error {
+	resolveProjectIdentifiers(&config, command, sys)
+	projectsScanned := triggerWhitesourceScan(command, config, sys)
+
+	if config.ProjectToken == "" {
+		// This state occurs when the project was created during the triggerWhitesourceScan step above,
+		resolveProjectIdentifiers(&config, command, sys) // we need to resolve the ProjectToken for pdf report download stage below
+	}
+
+	log.Entry().Info("Config.ProductVersion: " + config.ProductVersion)
+	log.Entry().Info("Config.ProjectToken: " + config.ProjectToken)
+	log.Entry().Infof("Number of projects scanned: %v", len(projectsScanned))
+
+	if config.Reporting {
+		for _, proj := range projectsScanned {
+			proj.Name = strings.Split(proj.Name, " - ")[0]
+			log.Entry().Infof("Attempting to download risk report for project name: %s", proj.Name)
+			downloadRiskReport(proj.Token, proj.Name, sys)
+		}
+	}
+	return nil
+}
+
+// translated from Groovy DSL: https://github.com/SAP/jenkins-library/blob/4c97231ff94d9f9fbeb1b58c3625c84b434a7de6/vars/whitesourceExecuteScan.groovy#L304
+func triggerWhitesourceScan(command *command.Command, config whitesourceExecuteScanOptions, sys whitesource.System) []whitesource.Project {
+	projectsScanned := []whitesource.Project{}
+	newProj := whitesource.Project{Name: config.ProjectName, Token: config.ProjectToken}
+	projectsScanned = append(projectsScanned, newProj)
+
+	switch config.ScanType {
+	case "npm":
+		executeNpmScan(config, command)
+		break
+
+	default:
+		// Auto generate a config file based on the current directory structure.
+		err := command.RunExecutable("java", "-jar", config.AgentFileName, "-d", ".", "-detect")
+		if err != nil {
+			log.Entry().WithError(err).Fatal("Failed to autogenerate Whitesource unified agent config file")
```

Comment:
@marcusholl @stippi2 in other steps we don't `panic` in functions but bubble the `err` up till `whitesourceExecuteScan`. Any specific reason for that?

Reply:
I am probably not aware of all the reasons, but one good reason is that it makes testing easier. The first step implementations used Fatal() at any level, and then later it was deemed better to bubble up errors and only exit in one place. I agree that this makes for better code in the end.